dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/DeadAssignementElimination.java

```diff
@@ -0,0 +1,87 @@
+import com.dat3m.dartagnan.program.event.core.utils.RegWriter;
+import static com.dat3m.dartagnan.program.event.Tag.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.sosy_lab.common.configuration.Configuration;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+
+// This is just Dead Store Elimination, but the use of the term "Store" can be confusing in our setting 
+public class DeadAssignementElimination implements ProgramProcessor {
+
+    private static final Logger logger = LogManager.getLogger(DeadAssignementElimination.class);
+    private Program program;
+
+    private DeadAssignementElimination() { }
+
+    public static DeadAssignementElimination newInstance() {
+        return new DeadAssignementElimination();
+    }
+
+    public static DeadAssignementElimination fromConfig(Configuration config) throws InvalidConfigurationException {
+        return newInstance();
+    }
+
+    @Override
+    public void run(Program program) {
+    	this.program = program;
+        logger.info("#Events before DSE: " + program.getEvents().size());
+
+        for (Thread t : program.getThreads()) {
+            eliminateDeadStores(t);
+            t.clearCache();
+        }
+        program.clearCache(false);
+
+        logger.info("#Events after DSE: " + program.getEvents().size());
+    }
+
+    private void eliminateDeadStores(Thread thread) {
+    	
+    	List<Event> removed = new ArrayList<>();
+    	
+    	// We traverse the events in inverse order and mark those that can be removed
+    	ArrayList<Event> sorted = new ArrayList<Event>(thread.getEvents());
+    	sorted.sort((Event e1, Event e2) -> e2.compareTo(e1));
+    	for(Event e : sorted) {
+    		// We cannot remove load events
+    		if(e.is(REG_WRITER) && !e.is(READ)) {
+    			RegWriter rw = (RegWriter)e;
+    			boolean notUsedByRegReader = e.getSuccessors().stream()
+    					.filter(s -> s instanceof RegReaderData && !removed.contains(s))
+    					.noneMatch(s -> ((RegReaderData)s).getDataRegs().contains(rw.getResultRegister()));
+    			boolean notUsedByMemEvent = e.getSuccessors().stream()
+    					.filter(s -> s instanceof MemEvent && !removed.contains(s))
+    					.noneMatch(s -> ((MemEvent)s).getAddress().getRegs().contains(rw.getResultRegister()));
+    			// We use register to pass the parameters to new threads
+    			boolean notUsedByOtherThread = program.getEvents().stream()
+    					.filter(o -> !o.getThread().equals(e.getThread()) && o instanceof RegReaderData && !removed.contains(o))
+    					.noneMatch(s -> ((RegReaderData)s).getDataRegs().contains(rw.getResultRegister()));
+    			boolean notUsedByAssertion = !program.getAss().getRegs().contains(rw.getResultRegister());
+				if(notUsedByRegReader && notUsedByMemEvent && notUsedByOtherThread && notUsedByAssertion) {
+    				removed.add(e);
+    			}
```

Comment:
Wouldn't it be easier to store a list of used registers?
Initially it is empty(*), but whenever you encounter a `MemEvent` that uses a register, you mark it as "used".
Whenever you encounter a local assignment, you check if it assigns to a used register and if not, you delete it. Otherwise, you add its dependencies to the used registers as well.

(*) To incorporate assertions and other threads, you can initially fill it with registers related to those.

Reply:
You also need to consider the case of `Jump`, since those also read registers. From an implementation point of view, you also need to split the case of `Load` and `Store` (and not just tread them as `MemEvent`) since for the former you need to add to "used" `getResultRegister` and for the later `getDataRegs`. I tried all this and the code becomes larger, thus I don't this what you propose is easier.

Reply:
I have pushed a new version that does what I proposed. I think it is a lot clear and the code size is pretty much identical (if it works correctly)